sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitExec.scala

```diff
@@ -0,0 +1,281 @@
+
+  override def increaseRank(): Unit = {
+    if (currentRankRow == null) {
+      currentRankRow = nextRow.copy()
+    } else {
+      if (ordering.compare(currentRankRow, nextRow) != 0) {
+        rank += 1
+        currentRankRow = nextRow.copy()
+      }
+    }
+  }
+}
+
+trait WindowIterator extends Iterator[InternalRow] {
+
+  def partitionSpec: Seq[Expression]
+
+  def output: Seq[Attribute]
+
+  def input: Iterator[InternalRow]
+
+  val grouping = UnsafeProjection.create(partitionSpec, output)
+
+  // Manage the stream and the grouping.
+  var nextRow: UnsafeRow = null
+  var nextGroup: UnsafeRow = null
+  var nextRowAvailable: Boolean = false
+  protected[this] def fetchNextRow(): Unit = {
+    nextRowAvailable = input.hasNext
+    if (nextRowAvailable) {
+      nextRow = input.next().asInstanceOf[UnsafeRow]
+      nextGroup = grouping(nextRow)
+    }
+  }
+  fetchNextRow()
+
+  var limitIterator: BaseLimitIterator = _
+
+  override final def hasNext: Boolean =
+    (limitIterator != null && limitIterator.hasNext) || nextRowAvailable
+
+  override final def next(): InternalRow = {
+    // Load the next partition if we need to.
+    if ((limitIterator == null || !limitIterator.hasNext) && nextRowAvailable) {
+      if (limitIterator != null) {
+        limitIterator.input.asInstanceOf[GroupIterator].skipRemainingRows()
+      }
+
+      if (nextRowAvailable) {
+        limitIterator = createLimitIterator(new GroupIterator())
+      } else {
+        // After skip remaining row in previous partition, all the input rows have been processed,
+        // so returns the last row directly.
+        return nextRow
+      }
+    }
+
+    if (limitIterator != null && limitIterator.hasNext) {
+      limitIterator.next()
+    } else {
+      throw new NoSuchElementException
+    }
+  }
+
+  class GroupIterator() extends Iterator[InternalRow] {
+    // Before we start to fetch new input rows, make a copy of nextGroup.
+    val currentGroup = nextGroup.copy()
+
+    def hasNext: Boolean = nextRowAvailable && nextGroup == currentGroup
+
+    def next(): InternalRow = if (nextRowAvailable && nextGroup == currentGroup) {
+      val currentRow = nextRow.copy()
+      fetchNextRow()
+      currentRow
+    } else {
+      throw new NoSuchElementException
+    }
+
+    def skipRemainingRows(): Unit = {
+      if (nextRowAvailable && nextGroup == currentGroup) {
+        // Skip all the remaining rows in this group
+        do {
+          fetchNextRow()
+        } while (nextRowAvailable && nextGroup == currentGroup)
+      }
+    }
+  }
+
+  protected def createLimitIterator(groupIterator: GroupIterator): BaseLimitIterator
+}
+
+case class SimpleGroupLimitIterator(
+    partitionSpec: Seq[Expression],
+    output: Seq[Attribute],
+    input: Iterator[InternalRow],
+    limit: Int) extends WindowIterator {
+
+  override def createLimitIterator(groupIterator: GroupIterator): BaseLimitIterator = {
+    SimpleLimitIterator(groupIterator, limit)
```

Comment:
for the worst case (one row per window group), this will instantiate one iterator each row, which can be bad for GC.